extension-workspace-reference.md

```diff
@@ -0,0 +1,111 @@
+	 * The name of the symbol containing the symbol that is being referenced.
+	 * For example:
+	 *
+	 * 	(Go) "Router"
+	 * 	(JS) "ReactMount"
+	 *
+	 */
+	containerName?: string;
+
+	/**
+	 * The file URI of the symbol that is being referenced. It may be either
+	 * a URI to the directory that contains the symbol OR the file that
+	 * contains the symbol, so long as the tuple (name, containerName, uri)
+	 * uniquely identifies the symbol.
+	 *
+	 * If `name` and `containerName` are not present, it indicates that this
+	 * URI location has been referenced but not a particular symbol. For
+	 * example, a Go import statement or a JS `require()` call, etc. which
+	 * references ('imports' or 'includes') a dependency but not a specific
+	 * symbol.
+	 */
+	uri: string;
+}
+```
+* error: code and message set in case an exception happens during the workspace reference request.
+
+## workspace/symbol query filters
+
+As part of this extension, we also require that language servers support two `workspace/symbol` query filters (specified via the `WorkspaceSymbolParams.query` field). These filters are:
+
+- `is:exported`
+  - When this filter is present in the query, only 'exported' symbols (according to language semantics) should be returned.
+- `dir:<directory>`
+  - When this filter is present in the query, only symbols that are located inside the specified directory path should be returned. Language servers should use this to optimize response times by limiting the work they must perform to only the specified directory.
+
+Note: The query filters are commonly space-separated. For example, a query will be issued for `dir:foo/bar Baz` where `foo/bar` is the directory path (relative to the workspace root) and `Baz` is a `containerName` from `ReferenceInformation` or `SymbolInformation`.
+
+## Design notes
+
+* The `ReferenceInformation` `(name, containerName, uri)` tuple being unique means that the user of this method can later invoke `workspace/symbol` or `textDocument/documentSymbol`, depending on whether or not the URI is a file or folder, and look for the same tuple in order to locate the exact position and kind of the symbol that is being referenced. Being able to lookup the exact position at a later time means that it is possible to e.g. jump to the reference.
+* `ReferenceInformation.uri` being a file OR folder URI (up to the language server implementation) is useful because while in some languages it is easier to define directory-scoped uniqueness (e.g. Go, where a package namespace aligns with a directory exactly), many other languages (e.g. Python, where each file holds a namespace) are not similar. The only requirement is that the `(name, containerName, uri)` tuple be unique and later resolvable via `workspace/symbol` (with the `is:exported` query) or `textDocument/documentSymbol`.
+* Clients may check the `ReferenceInformation.uri` for an extension (e.g. ending in `.go`) in order to determine whether or not the URI is a file or folder URI.
```

Comment:
The intention is to allow querying symbols via `textDocument/documentSymbol` as that would be more optimal if it is a file; and fallback to `workspace/symbol query="dir:..."` when it's a directory.

It might not be needed.. but it would be nice to give the option to LS devs. You're right it's a heuristic, though.